profiler/src/main/java/com/splunk/opentelemetry/logs/BatchingLogsProcessor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Splunk Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.logs;
+
+import static java.util.Objects.requireNonNull;
+
+import com.splunk.opentelemetry.profiler.util.HelpfulExecutors;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+
+/**
+ * A log processor that batches logs until a max number are buffered or a time limit has passed,
+ * which ever comes first.
+ */
+public class BatchingLogsProcessor implements LogsProcessor {
+
+  private static final int DEFAULT_MAX_BATCH_SIZE = 250;
+  private static final Duration DEFAULT_MAX_TIME_BETWEEN_BATCHES = Duration.ofSeconds(10);
+
+  private final int maxBatchSize;
+  private final Duration maxTimeBetweenBatches;
+  private final List<LogEntry> batch;
+  private final LogsExporter exporter;
+  private final ScheduledExecutorService executorService;
+  private final Object lock = new Object();
+  private WatchdogTimer watchdog;
+
+  public BatchingLogsProcessor(Builder builder) {
+    this.maxTimeBetweenBatches = builder.maxTimeBetweenBatches;
+    this.maxBatchSize = builder.maxBatchSize;
+    this.exporter = builder.batchAction;
+    this.executorService = builder.executorService;
+    batch = new ArrayList<>(maxBatchSize);
+  }
+
+  public void start() {
+    synchronized (lock) {
+      if (watchdog != null) {
+        throw new IllegalStateException("Already running");
+      }
+      watchdog = new WatchdogTimer(maxTimeBetweenBatches, this::syncDoExport, executorService);
+      watchdog.start();
+    }
+  }
+
+  void stop() {
+    synchronized (lock) {
+      if (watchdog == null) {
+        throw new IllegalStateException("Not running");
+      }
+      asyncDoExport();
+      watchdog.stop();
+      watchdog = null;
+    }
+  }
+
+  @Override
+  public void log(LogEntry log) {
+    synchronized (lock) {
+      batch.add(log);
+      if (batch.size() >= maxBatchSize) {
+        asyncDoExport();
+        watchdog.reset();
```

Comment:
I wonder if this `reset()` is really necessary. Would it be all that bad if the timed export just kept ticking along on it's normal schedule?

Reply:
Not _necessary_ per se, but it changes the behavior, and I like the "max size or timeout, whichever comes first" idea. If you remove this `reset()` then you could get an export with a tiny batch immediately after a full batch, which is a little unexpected/odd.